src/bsp/src/esp/src/SpiStm.cpp

```diff
@@ -0,0 +1,200 @@
+#include "SpiStm.h"
+#include "c-common/software_crc.h"
+#include "esp32/rom/crc.h"
+#include "hal/pin_map.h"
+#include "hal/spi_callbacks.h"
+#include <cstring>
+#include <driver/gpio.h>
+
+void task(void* instance) {
+    while (true) {
+        static_cast<SpiStm*>(instance)->execute();
+    }
+}
+
+SpiStm::SpiStm(ILogger& logger) : m_logger(logger) {
+    m_txState = transmitState::SENDING_HEADER;
+    m_rxState = receiveState::RECEIVING_HEADER;
+    m_inboundMessage = {0};
+    m_outboundMessage = {0};
+    m_inboundHeader = {0};
+    m_outboundHeader = {0};
+    m_isBusy = false;
+
+    setCallback(SpiStm::transactionCallback, this);
+
+    m_semaphore = xSemaphoreCreateBinaryStatic(&m_semaphoreBuffer);
+    xSemaphoreGive(m_semaphore);
+
+    m_taskHandle = xTaskCreateStatic(task, "esp_spi_driver_task", 4096, this, 1u,
+                                     m_stackData.data(), &m_stackBuffer);
+}
+
+bool SpiStm::send(const uint8_t* buffer, uint16_t length) {
+    bool retVal = false;
+
+    if (isBusy()) { // Not available
+    } else if (length >= STM_SPI_MAX_MESSAGE_LENGTH) { // Message too long
+        m_logger.log(LogLevel::Error,
+                     "StmEsp: Message length of %d is larger than maximum allowed of %d", length,
+                     STM_SPI_MAX_MESSAGE_LENGTH);
+    } else if (xSemaphoreTake(m_semaphore, 10) == pdTRUE) {
+        m_logger.log(LogLevel::Debug, "Sending message of length %d", length);
+        // Memcpy necessary to have buffer word-alligned for transfer
+        std::memcpy(m_outboundMessage.m_data.data(), buffer, length);
+        // Padding with 0 up to a word-alligned boundary
+        for (uint8_t i = 0; i < (length % 4); i++) {
+            m_outboundMessage.m_data[length] = 0;
+            length++;
+        }
+        // Appending CRC32
+        *(uint32_t*)(m_outboundMessage.m_data.data() + length) =
+            calculateCRC32_software(buffer, length);
+        ;
+        length += CRC32_SIZE;
+        m_outboundMessage.m_sizeBytes = length;
+        m_isBusy = true;
+        notifyMaster();
+        xSemaphoreGive(m_semaphore);
```

Comment:
What is critical here? the notifyMaster?

Reply:
Removed semaphore, proved to be useless